### ISSUE TITLE
How the auto commit is working and overall recommendations on fine tuning consumer

### ISSUE THREAD

**Issue by cubbuk** [tulios/kafkajs#707]
I am reading the section on autocommit and I wonder how does offset committing work out of the box? 

https://kafka.js.org/docs/consuming#a-name-auto-commit-a-autocommit

As far as I understand by default auto committing is enabled however `autoCommitInterval` and `autoCommitThreshold` are set to null. So with default values do we still get auto committing and if so according to which rules it works?

We have a long running process (can take up to 60 seconds time to time) and I see that due to rebalancing we sometimes end up in a loop where we keep processing the same records but cannot commit the offset. And after a rebalance the consumer starts from the earlier offset.

For avoiding it we increased the session timeout to 90000 and heartbeat interval to 45000. Interesting bit is we can see that heartbeat keeps happening but offset does not get committed.

So I would really appreciate any recommendations on how to fine tune our consumer for such a process.

**Comment by Nevon** [comment]
The answer depends on whether you are using the `eachBatch` or `eachMessage` interface.

Let's take `eachMessage` first. In that scenario, the userland code yields back to us in between each message, and we take that opportunity to call `autoCommitIfNecessary`. That function will check if either `autoCommitInterval` or `autoCommitThreshold` have been exceeded, and if so, commit. By default, since both are null, it will not commit. You can see this here: https://github.com/tulios/kafkajs/blob/9b65843638e73dd6d3360a9b5faf45680a184826/src/consumer/runner.js#L158-L184

What you can also see is that if there is an error processing your message, we commit offsets (regardless of any autocommit setting).

Now let's take a look at `eachBatch`. In `eachBatch` the userland code only yields back control to us in between batches, so we can't do any committing while you're processing the batch. However, we provide your `eachBatch` function with a function called `commitOffsetsIfNecessary`, which you can call to auto commit yourself whenever it makes sense for you (such as perhaps in between each message): https://github.com/tulios/kafkajs/blob/9b65843638e73dd6d3360a9b5faf45680a184826/src/consumer/runner.js#L217-L227

As you can see, this just delegates to `autoCommitIfNecessary`, the same method called from `eachMessage`, so it behaves the same as I described for `eachMessage`.

In case you throw from `eachBatch`, we also call `autoCommitOffsets`, assuming you haven't disabled it. The intention is that we should commit any resolved offsets so that you don't have to reprocess them, if possible. https://github.com/tulios/kafkajs/blob/9b65843638e73dd6d3360a9b5faf45680a184826/src/consumer/runner.js#L232-L253

After successfully processing all batches we received from this fetch, regardless of whether you are using `eachMessage` or `eachBatch`, we finish off by committing all resolved offsets, so that the next fetch can happen: https://github.com/tulios/kafkajs/blob/9b65843638e73dd6d3360a9b5faf45680a184826/src/consumer/runner.js#L360-L370

So basically, if you're using `eachMessage`, we'll try to commit offsets after each message, and you can control the frequency with `autoCommitThreshold` and/or `autoCommitInterval`. If you're using `eachBatch`, you have to invoke `commitOffsetsIfNecessary` yourself whenever you think it makes sense to potentially commit offsets. If you haven't configured `autoCommitThreshold` or `autoCommitInterval`, `commitOffsetsIfNecessary` is essentially a noop.

**Comment by cubbuk** [comment]
Thanks a lot @Nevon.

So in that case for `eachMessage` I assume library commits after each message right?

We use `eachMessage` to process messages. I think the issue we had was related to happening around getting unnecessary rebalances due to not closing our consumers in case of a new deployment. 

We actually listen to `SIGTERM` and `SIGINT` but since we were running our process as npm script, those signals were not handled properly. After switching to running the processes directly through `node` consumers started to disconnect themselves properly during deployments and we stopped getting this kind of problem.

We also started setting `autoCommitInterval` to 10 and maybe that helped too not sure. 

Thanks again for the through explanation :+1:

**Comment by Nevon** [comment]
> So in that case for eachMessage I assume library commits after each message right?

No, it will basically check if the `autoCommitInterval` or `autoCommitThreshold` limits have been reached, and if one of them has, it will commit.

There are two concepts to understand: `resolve`  and `commit`.

`resolve` is just an in memory check saying "I have processed offset X". This is extremely cheap (literally just setting a value), and is done after each message when using `eachMessage`. This value is used when making the next fetch request ("I want messages from <resolved offset>")

`commit` means to actually make a network request to the broker, letting it know that the consumer group has processed a certain offset. This is relatively expensive, so we want to avoid doing it unless necessary.

In the `eachMessage` case, we normally resolve after each message, and then commit once all messages have been processed OR `eachMessage` throws an error.

**Comment by cubbuk** [comment]
I see, its much clear now. So I think us setting `autoCommitThreshold` to 10 made it work for us. Because I am pretty sure we were processing the same message as it was taking longer than the session period and the processed messages were not committed.

In that case I think the flag name `autoCommit` is a bit misleading as I would expect it to auto commit without doing anything. However with default configuration where autoCommit is set to true and `autoCommitInterval` and `autoCommitThreshold` is not set, this actually auto commit the batch but not the message.

How about logging a warning when `autoCommit` is set but none of the threshold values are set so that the user would understand the behavior easier. `Auto Commit is enabled without any threshold value, messages will only get committed after processing each batch. For committing each messages after processing it, you should set `autoCommitThreshold` to 1. Beware this can impact the performance as committing offset is an expensive operation`

**Comment by Nevon** [comment]
> How about logging a warning when autoCommit is set but none of the threshold values are set so that the user would understand the behavior easier.

I'm not so sure. The reason having to do with historical usage of KafkaJS, and the expectations of existing users.

The original behavior of KafkaJS was the same as `autoCommit: true`, `autoCommitInterval: null`, `autoCommitThreshold: null` - we introduced it that way so as to not break backwards compatibility. So that's really the expected behavior for existing users, and it is the best configuration _in most cases_.

`autoCommit` was introduced as a way to disable committing in between each batch. So the vast majority of users don't specify `autoCommit` at all, and just commit in between batches, and then some users that handle offsets outside of Kafka use the option to disable committing.

`autoCommitInterval` and `autoCommitThreshold` were introduced after that because we wanted a simple way for users to commit periodically, without having to have any complex logic on their side. Looking back, I'm not sure this was the right move. For the vast, vast majority of users, committing in between each batch, on error or on shutdown should be all that's needed.

So I would be hesitant to introduce any kind of notice or warning for the configuration that is actually what I would consider the ideal for most use-cases. Very few users should need to configure `autoCommitInterval` and `autoCommitThreshold`.

**Comment by cubbuk** [comment]
I am definitely not a good example for representing experienced kafka users as I recently started using it. 

And I was a bit confused with the default configuration but that's most probably related to my lack of experience with kafka and I agree that the library should not be bothering with such cases but maybe a warning on docs for beginners can help to avoid such a pitfall. Because regardless of experience, I think the default behavior you get out of the box is a bit unexpected, as the default configuration of `autoCommit`, `autoCommitInterval` and `autoCommitThreshold` is a bit contradicting. 

Maybe its just me so thanks again for clarifying things it was really helpful :+1:

**Comment by Nevon** [comment]
We should _definitely_ cater for the first time user as well, and as you point out, the docs seem like a good place to do so. Reading through the existing docs now, though, they seem to explain what I just explained pretty well - but that could just be because I already know how this stuff works. Is there anything in there that you think doesn't make sense? https://kafka.js.org/docs/consuming#a-name-auto-commit-a-autocommit

> The messages are always fetched in batches from Kafka, even when using the eachMessage handler. All resolved offsets will be committed to Kafka after processing the whole batch.

Describes the default behavior.

> Committing offsets periodically during a batch allows the consumer to recover from group rebalances, stale metadata and other issues before it has completed the entire batch. However, committing more often increases network traffic and slows down processing. 

Explains why the default behavior might want to be overriden.

> Auto-commit offers more flexibility when committing offsets; there are two flavors available
> [...]
> Having both flavors at the same time is also possible, the consumer will commit the offsets if any of the use cases (interval or number of messages) happens.

Explains the options available and what happens if you combine the two.

> `autoCommit`: Advanced option to disable auto committing altogether. Instead, you can manually commit offsets. Default: `true`

Explains how to disable automatic commits entirely for more advanced use-cases.

**Comment by cubbuk** [comment]
I think this bit was where I missed the point ` All resolved offsets will be committed to Kafka after processing the whole batch.` As I didn't know the resolve offset concept, I wasn't able to understand what it actually means. I think adding smtg like to that explanation would give a better guidance

> Since kafka sends data in batches, this means when `autoCommit` is enabled without setting `autoCommitThreshold` or `autoCommitInterval`, `kafkajs` will only commit offsets after the whole batch is processed by your consumer. If you want to commit offsets after each message, you should set `autoCommitThreshold` to 1. Beware that committing more often increases network traffic and slows down processing.

But sincerely after your explanation, what we have on the documentation sounds pretty clear also, so it might be an overkill to add the line above :)

**Comment by matt-dowds** [comment]
Just to chip in on this as another new user who has also been getting confused by this behaviour:

The thing that caught me out was the fact that by default there is no auto-commit interval. Rereading the docs I can see that this is explained, but given that this is different behaviour than node-rdkafka and the Java Kafka client it might be worth making that difference more explicit for new users.

**Comment by snooyen** [comment]
@tulios @Nevon Sorry to revive an old issue, but was wondering if setting lower values for autoCommitInterval (autocommit occurs at higher frequency) could lead to possible data loss in the event of a crash? Something like what is described [here](https://dev.to/yukooshima/understanding-and-mitigating-message-loss-in-apache-kafka-28jp#:~:text=Consider%20a%20scenario%20where%20the,on%20to%20the%20next%20batch.)?

Or is the resolvedOffsets attribute of the offset manager supposed to mitigate that?
